doc/specs/#4999 - Improved keyboard handling in Conpty.md

```diff
@@ -0,0 +1,559 @@
+
+### Inspiration
+
+The design we've settled upon is one that's highly inspired by a few precedents:
+* `Application Cursor Keys (DECCKM)` is a long-supported VT sequence which a
+  client application can use to request a different input format from the
+  Terminal. This is the DECSET sequence `^[[?1h`/`^[[?1l` (for enable/disable,
+  respectively). This changes the sequences sent by keys like the Arrow keys
+  from a sequence like `^[[A` to `^[OA` instead.
+* The `kitty` terminal emulator uses a similar DECSET sequence for enabling
+  their own input format, which they call ["full mode"]. Similar to DECCKM, this
+  changes the format of the sequences that the terminal should send for keyboard
+  input. Their "full mode" contains much more information when keys are pressed
+  or released (though, less than a full `INPUT_RECORD` worth of data). Instead
+  of input being sent to the client as a CSI or SS3 sequence, this `kitty` mode
+  uses "Application Program-Command" (or "APC") sequences , prefixed with `^[_`.
+* [iTerm2](https://www.iterm2.com/documentation-escape-codes.html) has a region
+  of OSC's that they've carved for themselves all starting with the same initial
+  parameter, `1337`. They then have a number of commands that all use the second
+  parameter to indicate what command specific to iTerm2 they're actually
+  implementing.
+
+### Requesting `win32-input-mode`
+
+An application can request `win32-input-mode` with the following private mode sequence:
+
+```
+  ^[ [ ? 9001 h/l
+              l: Disable win32-input-mode
+              h: Enable win32-input-mode
+```
+
+Private mode `9001` seems unused according to the [xterm
+documentation](https://invisible-island.net/xterm/ctlseqs/ctlseqs.html). This is
+stylistically similar to how `DECKPM`, `DECCKM`, and `kitty`'s ["full mode"] are
+enabled.
+
+> 👉 NOTE: an earlier draft of this spec used an OSC sequence for enabling these
+> sequences. This was abandoned in favor of the more stylistically consistent
+> private mode params proposed above. Additionally, if implemented as a private
+> mode, then a client app could query if this setting was set with `DECRQM`
+
+When a terminal receives a `^[[?9001h` sequence, they should switch into
+`win32-input-mode`. In `win32-input-mode`, the terminal will send keyboard input
+to the connected client application in the following format:
+
+### `win32-input-mode` sequences
+
+The `KEY_EVENT_RECORD` portion of an input record (the part that's important for
+us to encode in this feature) is defined as the following:
+
+```c++
+typedef struct _KEY_EVENT_RECORD {
+  BOOL  bKeyDown;
+  WORD  wRepeatCount;
+  WORD  wVirtualKeyCode;
+  WORD  wVirtualScanCode;
+  union {
+    WCHAR UnicodeChar;
+    CHAR  AsciiChar;
+  } uChar;
+  DWORD dwControlKeyState;
+} KEY_EVENT_RECORD;
+```
+
+To encode all of this information, I propose the following sequence. This is a
+CSI sequence with a final terminator character of `_`. This character appears to

```

Comment:
ECMA-48 clause 5.4 lists `_` (05/15) as an unallocated bit combination that is "reserved for future standardization and shall not be used." Did you consider using a Final Byte from the range 07/00 to 07/14, which is available for private use? I see VT510 uses many of those already; perhaps there aren't any left if you want to avoid conflicts with everything and not use Intermediate Bytes.

Reply:
That's a good point. Something from the private range would have been more appropriate. But the only character I know that _might_ be unused is `{` (07/11). Since this sequence is only used internally, it's probably not too late to change if anyone feels strongly about it.

Reply:
<https://vt100.net/docs/vt510-rm/chapter4.html> lists two functions with `{` as the Final Byte, but each of those includes an Intermediate Byte.

- Selective Erase Rectangular Area: CSI _Pt_;_Pl_;_Pb_;_Pr_ $ {
- Macro Space Report: CSI _Pn_ * {

Reply:
Yeah, there are a quite a few sequences using it with an intermediate. I just meant as a final byte by itself. If we're willing to use an intermediate, there are probably a lot more options available.

Reply:
I just remembered we discussed this before, and as an input sequence, we found nearly every valid final char was used, so even `{` isn't technically an option if we want to avoid all conflicts. But since this sequence is only used if the win32 input mode is explicitly enabled, we decided it didn't really matter. I suspect nobody will care about it not being private for the same reason.